Feature/initial implementation

CarlHembrough opened the pull request:
Initial implementation of a dataset API client. Covers only the endpoints currently required by Zebedee.

CarlHembrough committed:
Migrate existing code from zebedee.

CarlHembrough committed:
Add travis build file

CarlHembrough committed:
Add create dataset. Add tests. Refactoring.

CarlHembrough committed:
Fixing tests. Junit5 doesn't seem to like member vars

CarlHembrough committed:
Remove dataset not found check from createDataset

CarlHembrough committed:
Add URI field to dataset model.

CarlHembrough committed:
Simplify URI creation by using resolve instead of builder

Matt-Guest commented on pom.xml:
```diff
@@ -8,5 +8,73 @@
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.9.0</version>
```
There now version 3.9.1 from 21/2/2018

CarlHembrough replied:
done

Matt-Guest commented on pom.xml:
```diff
@@ -8,5 +8,73 @@
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.3</version>
```
Can this be pushed up to 3.7?

CarlHembrough replied:
done

Matt-Guest commented on src/main/java/dp/api/dataset/DatasetAPIClient.java:
```diff
@@ -0,0 +1,378 @@
+/**
+ * HTTP client for the dataset API.
+ */
+public class DatasetAPIClient implements DatasetClient {
```
Can this class extend closable as well? as CloseableHTTPClient is a internal resource.

CarlHembrough replied:
done

Matt-Guest commented on src/main/java/dp/api/dataset/DatasetAPIClient.java:
```diff
@@ -0,0 +1,378 @@
+                httpRequest.getURI());
+    }
+
+    public static void main(String[] args) throws Exception {
```
Can the main method be moved into a Example class? As anything that imports the client will also take this main.

CarlHembrough replied:
Removed main method

Matt-Guest commented on src/main/java/dp/api/dataset/DatasetAPIClient.java:
```diff
@@ -0,0 +1,378 @@
+        dataset.setTitle("123");
+
+        //System.out.println(json.writeValueAsString(dataset));
+        //datasetAPIClient.createDataset("123", dataset);
```
Commented out code here

CarlHembrough replied:
Removed main method

Matt-Guest commented on src/main/java/dp/api/dataset/model/Dataset.java:
```diff
@@ -0,0 +1,83 @@
+        this.uri = uri;
+    }
+
+    public static void main(String[] args) throws IOException {
```
Is this needed?

CarlHembrough replied:
No I have removed it 👍

CarlHembrough committed:
Add deleteDataset

Matt-Guest commented on src/main/java/dp/api/dataset/DatasetAPIClient.java:
```diff
@@ -0,0 +1,378 @@
+    private final URI datasetAPIURL;
+    private final String datasetAPIAuthToken;
+
+    private static final String authTokenHeaderName = "internal-token";
```
This should be "Internal-Token"

CarlHembrough replied:
Fixed

Matt-Guest commented on src/test/java/dp/api/dataset/DatasetAPIClientTest.java:
```diff
@@ -0,0 +1,647 @@
+
+    private static final ObjectMapper json = new ObjectMapper();
+
+    private static final String authTokenHeaderName = "internal-token";
```
This should be "Internal-Token"

CarlHembrough replied:
Fixed

CarlHembrough committed:
Add transitive dependencies for jackson.

https://stackoverflow.com/questions/27746750/exception-in-thread-main-java-lang-nosuchmethoderror-com-fasterxml-jackson-co

CarlHembrough committed:
Add json include annotations for all models

CarlHembrough committed:
Add @JsonIgnoreProperties annotation to models

CarlHembrough committed:
Ignore unknown properties when serializing JSON

CarlHembrough committed:
Add JsonProperty annotations for dataset state enum

CarlHembrough committed:
Add test for json serialisation handling unknown fields

CarlHembrough committed:
Add taxonomy link to dataset model

CarlHembrough committed:
throw specific exceptions for forbidden / unauthorised responses

CarlHembrough committed:
Revert deleted file

CarlHembrough committed:
Update dependency versions

CarlHembrough committed:
Implement closable on dataset API client

CarlHembrough committed:
Fix casing on auth header + remove redundant code

CarlHembrough committed:
Make DatasetClient interface extend closeable

CarlHembrough committed:
Remove unused imports
